### PR TITLE
Convert Address to Kotlin

### LINKED ISSUE #4765: Migrate Address to Kotlin
I'd like to take a stab at this! 
Ref #4723

### PR DESCRIPTION
**Pull request by shaishavgandhi05**
Resolves #4765 
The build will fail right now since `OkHttpClient` doesn't recognize `internal` modifier on `equalsNonHost`. Will wait for #4748 for that. 

Also made the properties private as advised in #4782.

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/Address.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+/**
+ * A specification for a connection to an origin server. For simple connections, this is the
+ * server's hostname and port. If an explicit proxy is requested (or [no][Proxy.NO_PROXY] is explicitly requested), this also includes that proxy information. For secure
+ * connections the address also includes the SSL socket factory, hostname verifier, and certificate
+ * pinner.
+ *
+ * HTTP requests that share the same `Address` may also share the same [Connection].
+ */
+class Address(
```
Comment (shaishavgandhi05) [inline okhttp/src/main/java/okhttp3/Address.kt]: Thought about making this a `data` class but then we'd have to override `hashCode`, `equals` and `toString()` anyway. 

Thoughts?

File: okhttp/src/main/java/okhttp3/Address.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import java.net.ProxySelector
+import java.util.Objects
+import javax.net.SocketFactory
+import javax.net.ssl.HostnameVerifier
+import javax.net.ssl.SSLSocketFactory
+import okhttp3.internal.Util
+
+/**
+ * A specification for a connection to an origin server. For simple connections, this is the
+ * server's hostname and port. If an explicit proxy is requested (or [no][Proxy.NO_PROXY] is explicitly requested), this also includes that proxy information. For secure
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/Address.kt]: no -> no proxy

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Address.kt]: Wrap these at 100 columns.

File: okhttp/src/main/java/okhttp3/Address.kt (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Address.kt]: As-is for now please.

File: okhttp/src/main/java/okhttp3/Address.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+   * this URL are always empty, since they are not significant for planning a route.
+   */
+  fun url() = url
+
+  /** Returns the service that will be used to resolve IP addresses for hostnames. */
+  fun dns() = dns
+
+  /** Returns the socket factory for new connections. */
+  fun socketFactory(): SocketFactory {
+    return socketFactory
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/Address.kt]: Expression function?

File: okhttp/src/main/java/okhttp3/Address.kt [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        && this.proxySelector == that.proxySelector
+        && this.proxy == that.proxy
+        && this.sslSocketFactory == that.sslSocketFactory
+        && this.hostnameVerifier == that.hostnameVerifier
+        && this.certificatePinner == that.certificatePinner
+        && this.url().port() == that.url().port())
+  }
+
+  override fun toString(): String {
+    val result = StringBuilder()
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/Address.kt]: buildString ?

File: okhttp/src/main/java/okhttp3/Address.kt [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+      .scheme(if (sslSocketFactory != null) "https" else "http")
+      .host(uriHost)
+      .port(uriPort)
+      .build()
+
+  /**
+   * The protocols the client supports. This method always returns a non-null list that
+   * contains minimally [Protocol.HTTP_1_1].
+   */
+  private val protocols: List<Protocol> = Util.immutableList(protocols)
```
Comment (oldergod) [inline okhttp/src/main/java/okhttp3/Address.kt]: @swankjesse do we need all those `Util.immutableList` in Kotlin?

File: okhttp/src/main/java/okhttp3/Address.kt [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  /** Returns the hostname verifier, or null if this is not an HTTPS address. */
+  fun hostnameVerifier() = hostnameVerifier
+
+  /** Returns this address's certificate pinner, or null if this is not an HTTPS address. */
+  fun certificatePinner() = certificatePinner
+
+  override fun equals(other: Any?): Boolean {
+    return (other is Address
+        && url == other.url
+        && equalsNonHost((other as Address?)!!))
```
Comment (oldergod) [inline okhttp/src/main/java/okhttp3/Address.kt]: I believe you can cast to the non-nullable type `other as Address`

Comment (JakeWharton) [reply okhttp/src/main/java/okhttp3/Address.kt]: The explicit cast shouldn't even be needed. Smart cast should take care of it.

File: okhttp/src/main/java/okhttp3/Address.kt (hunk 2 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Address.kt]: Fixed in e751fc152f6b74804a016a119e8fbc64ae85142f

File: okhttp/src/main/java/okhttp3/Address.kt (hunk 3 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Address.kt]: Done in e751fc152f6b74804a016a119e8fbc64ae85142f

File: okhttp/src/main/java/okhttp3/Address.kt (hunk 6 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Address.kt]: Yup, good point! Done in e751fc152f6b74804a016a119e8fbc64ae85142f

File: okhttp/src/main/java/okhttp3/Address.kt (hunk 4 above)
Comment (shaishavgandhi05) [reply okhttp/src/main/java/okhttp3/Address.kt]: Done in e751fc152f6b74804a016a119e8fbc64ae85142f

File: okhttp/build.gradle [hunk 7]
```diff
@@ -90,6 +90,18 @@ task japicmp(type: me.champeau.gradle.japicmp.JapicmpTask, dependsOn: 'jar') {
...
+      'okhttp3.Address#dns()',
+      'okhttp3.Address#hostnameVerifier()',
+      'okhttp3.Address#protocols()',
+      'okhttp3.Address#proxy()',
+      'okhttp3.Address#proxyAuthenticator()',
+      'okhttp3.Address#proxySelector()',
+      'okhttp3.Address#sslSocketFactory()',
+      'okhttp3.Address#url()',
+      'okhttp3.Address#socketFactory()',
+      'okhttp3.TlsVersion#javaName()',
```
Comment (swankjesse) [inline okhttp/build.gradle]: tiny merge conflict here; this is repeated

File: okhttp/src/main/java/okhttp3/Address.kt (hunk 5 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Address.kt]: Probably. Kotlin’s immutable list isn’t enforced very strongly.

File: okhttp/src/main/java/okhttp3/Address.kt [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        && this.proxySelector == that.proxySelector
+        && this.proxy == that.proxy
+        && this.sslSocketFactory == that.sslSocketFactory
+        && this.hostnameVerifier == that.hostnameVerifier
+        && this.certificatePinner == that.certificatePinner
+        && this.url().port() == that.url().port())
+  }
+
+  override fun toString(): String {
+    return buildString {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/Address.kt]: 👍🏻

**Comment by swankjesse** [comment]
Uh oh. The fix I’ve been using is to add a helper method to `InternalKt.kt`.
```
> Task :okhttp:compileJava
/home/travis/build/square/okhttp/okhttp/src/main/java/okhttp3/OkHttpClient.java:145: error: cannot find symbol
        return a.equalsNonHost(b);
                ^
  symbol:   method equalsNonHost(Address)
  location: variable a of type Address
Note: /home/travis/build/square/okhttp/okhttp/src/main/java/okhttp3/OkHttpClient.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
1 error
```

File: okhttp/build.gradle (hunk 7 above)
Comment (shaishavgandhi05) [reply okhttp/build.gradle]: Woops. Fixed in bacfcb2e26b952bbd5f34c82d137a73646998509

**Pull request merged**